File: go-page-server/core/spider_detector.go
// Package core contains the core business logic
package core

import (
	"sync"
	"sync/atomic"
	"time"

	lru "github.com/hashicorp/golang-lru/v2"
	"github.com/rs/zerolog/log"
	"go-page-server/models"
)

// SpiderInfo holds spider configuration (kept for backward compatibility)
type SpiderInfo struct {
	Type       string
	Name       string
	DNSDomains []string
}

// cacheEntry represents a cached detection result with TTL
type cacheEntry struct {
	spiderType string
	expireAt   time.Time
}

// SpiderDetector detects search engine spiders by User-Agent
type SpiderDetector struct {
	configLoader *SpiderConfigLoader
	cache        *lru.Cache[string, *cacheEntry]
	cacheEnabled bool
	cacheTTL     time.Duration
	cacheHits    int64
	cacheMisses  int64
	mu           sync.RWMutex
}

// NewSpiderDetector creates a new spider detector (backward compatible version)
func NewSpiderDetector() *SpiderDetector {
	// Try to load from default config path
	configPath := DefaultSpiderConfigPath()
	detector, err := NewSpiderDetectorWithConfig(configPath)
	if err != nil {
		log.Warn().Err(err).Msg("Failed to load spider config, using fallback hardcoded patterns")
		return newSpiderDetectorFallback()
	}
	return detector
}

// NewSpiderDetectorWithConfig creates a new spider detector with specified config path
func NewSpiderDetectorWithConfig(configPath string) (*SpiderDetector, error) {
	loader, err := NewSpiderConfigLoader(configPath)
	if err != nil {
		return nil, err
	}

	config := loader.GetConfig()

	sd := &SpiderDetector{
		configLoader: loader,
		cacheEnabled: config.Cache.Enabled,
		cacheTTL:     time.Duration(config.Cache.TTLSeconds) * time.Second,
	}

	// Initialize LRU cache if enabled
	if config.Cache.Enabled {
		maxSize := config.Cache.MaxSize
		if maxSize <= 0 {
			maxSize = 10000 // default
		}
		cache, err := lru.New[string, *cacheEntry](maxSize)
		if err != nil {
			return nil, err
		}
		sd.cache = cache
	}

	// Set up hot-reload callback
	loader.OnChange(func(newConfig *SpiderConfig, rules []*CompiledSpiderRule) {
		sd.onConfigChange(newConfig)
	})

	return sd, nil
}

// onConfigChange handles configuration changes
func (sd *SpiderDetector) onConfigChange(newConfig *SpiderConfig) {
	sd.mu.Lock()
	defer sd.mu.Unlock()

	// Update cache settings
	sd.cacheEnabled = newConfig.Cache.Enabled
	sd.cacheTTL = time.Duration(newConfig.Cache.TTLSeconds) * time.Second

	// Recreate cache if size changed
	if sd.cacheEnabled && sd.cache != nil {
		newSize := newConfig.Cache.MaxSize
		if newSize <= 0 {
			newSize = 10000
		}
		// Only recreate if size changed significantly
		if newSize != sd.cache.Len() {
			newCache, err := lru.New[string, *cacheEntry](newSize)
			if err != nil {
				log.Error().Err(err).Msg("Failed to recreate cache on config change")
				return
			}
			sd.cache = newCache
			log.Info().Int("new_size", newSize).Msg("Cache recreated with new size")
		}
	}

	// Clear cache on config change to ensure new rules take effect
	if sd.cache != nil {
		sd.cache.Purge()
		log.Info().Msg("Cache purged due to configuration change")
	}
}

// StartWatching starts watching for configuration changes
func (sd *SpiderDetector) StartWatching() error {
	if sd.configLoader == nil {
		return nil
	}
	return sd.configLoader.WatchChanges()
}

// StopWatching stops watching for configuration changes
func (sd *SpiderDetector) StopWatching() {
	if sd.configLoader != nil {
		sd.configLoader.Stop()
	}
}

// Detect detects if the User-Agent belongs to a spider
func (sd *SpiderDetector) Detect(userAgent string) *models.DetectionResult {
	if userAgent == "" {
		return &models.DetectionResult{
			IsSpider:  false,
			UserAgent: userAgent,
		}
	}

	// Check cache first (if enabled)
	sd.mu.RLock()
	cacheEnabled := sd.cacheEnabled
	cache := sd.cache
	cacheTTL := sd.cacheTTL
	sd.mu.RUnlock()

	if cacheEnabled && cache != nil {
		if entry, ok := cache.Get(userAgent); ok {
			// Check if entry has expired
			if time.Now().Before(entry.expireAt) {
				atomic.AddInt64(&sd.cacheHits, 1)
				if entry.spiderType == "" {
					return &models.DetectionResult{
						IsSpider:  false,
						UserAgent: userAgent,
					}
				}
				rule := sd.configLoader.GetRuleByType(entry.spiderType)
				if rule != nil {
					return &models.DetectionResult{
						IsSpider:   true,
						SpiderType: entry.spiderType,
						SpiderName: rule.Name,
						UserAgent:  userAgent,
					}
				}
			}
			// Entry expired, remove it
			cache.Remove(userAgent)
		}
	}

	atomic.AddInt64(&sd.cacheMisses, 1)

	// Get compiled rules from config loader
	rules := sd.configLoader.GetCompiledRules()

	// Match against patterns
	for _, rule := range rules {
		if !rule.Enabled {
			continue
		}
		for _, pattern := range rule.Patterns {
			if pattern.MatchString(userAgent) {
				// Cache the result
				if cacheEnabled && cache != nil {
					cache.Add(userAgent, &cacheEntry{
						spiderType: rule.Type,
						expireAt:   time.Now().Add(cacheTTL),
					})
				}

				return &models.DetectionResult{
					IsSpider:   true,
					SpiderType: rule.Type,
					SpiderName: rule.Name,
					UserAgent:  userAgent,
				}
			}
		}
	}

	// Not a spider, cache negative result
	if cacheEnabled && cache != nil {
		cache.Add(userAgent, &cacheEntry{
			spiderType: "",
			expireAt:   time.Now().Add(cacheTTL),
		})
	}

	return &models.DetectionResult{
		IsSpider:  false,
		UserAgent: userAgent,
	}
}

// IsSpider is a quick check if the UA is a spider
func (sd *SpiderDetector) IsSpider(userAgent string) bool {
	result := sd.Detect(userAgent)
	return result.IsSpider
}

// GetStats returns cache statistics
func (sd *SpiderDetector) GetStats() map[string]interface{} {
	sd.mu.RLock()
	cache := sd.cache
	sd.mu.RUnlock()

	cacheSize := 0
	if cache != nil {
		cacheSize = cache.Len()
	}

	return map[string]interface{}{
		"cache_size":   cacheSize,
		"cache_hits":   atomic.LoadInt64(&sd.cacheHits),
		"cache_misses": atomic.LoadInt64(&sd.cacheMisses),
	}
}

// GetSpiderInfo returns information about a specific spider type
func (sd *SpiderDetector) GetSpiderInfo(spiderType string) *SpiderInfo {
	if sd.configLoader == nil {
		return nil
	}
	rule := sd.configLoader.GetRuleByType(spiderType)
	if rule == nil {
		return nil
	}
	return &SpiderInfo{
		Type:       rule.Type,
		Name:       rule.Name,
		DNSDomains: rule.DNSDomains,
	}
}

// GetAllSpiderTypes returns all configured spider types
func (sd *SpiderDetector) GetAllSpiderTypes() []string {
	if sd.configLoader == nil {
		return nil
	}
	config := sd.configLoader.GetConfig()
	if config == nil {
		return nil
	}
	types := make([]string, 0, len(config.Spiders))
	for spiderType := range config.Spiders {
		types = append(types, spiderType)
	}
	return types
}

// newSpiderDetectorFallback creates a fallback detector with hardcoded patterns
// This is used when configuration file is not available
func newSpiderDetectorFallback() *SpiderDetector {
	sd := &SpiderDetector{
		cacheEnabled: true,
		cacheTTL:     time.Hour,
	}

	// Create a simple in-memory cache
	cache, _ := lru.New[string, *cacheEntry](10000)
	sd.cache = cache

	// Create a minimal config loader with hardcoded rules
	sd.configLoader = &SpiderConfigLoader{
		config: &SpiderConfig{
			Cache: SpiderCacheConfig{
				Enabled:    true,
				MaxSize:    10000,
				TTLSeconds: 3600,
			},
			Spiders: getDefaultSpiderRules(),
		},
		rulesByType: make(map[string]*CompiledSpiderRule),
	}

	// Compile hardcoded rules
	rules, rulesByType, _ := sd.configLoader.compileRules(sd.configLoader.config)
	sd.configLoader.compiledRules = rules
	sd.configLoader.rulesByType = rulesByType

	log.Info().Msg("Spider detector initialized with fallback hardcoded patterns")
	return sd
}

// getDefaultSpiderRules returns the default hardcoded spider rules
func getDefaultSpiderRules() map[string]SpiderRule {
	return map[string]SpiderRule{
		"baidu": {
			Name:       "百度蜘蛛",
			Enabled:    true,
			Patterns:   []string{`(?i)Baiduspider`, `(?i)Baidu-YunGuanCe`},
			DNSDomains: []string{"baidu.com", "baidu.jp"},
		},
		"google": {
			Name:       "谷歌蜘蛛",
			Enabled:    true,
			Patterns:   []string{`(?i)Googlebot`, `(?i)Google-InspectionTool`, `(?i)Mediapartners-Google`},
			DNSDomains: []string{"googlebot.com", "google.com"},
		},
		"bing": {
			Name:       "必应蜘蛛",
			Enabled:    true,
			Patterns:   []string{`(?i)bingbot`, `(?i)msnbot`, `(?i)BingPreview`},
			DNSDomains: []string{"search.msn.com"},
		},
		"sogou": {
			Name:       "搜狗蜘蛛",
			Enabled:    true,
			Patterns:   []string{`(?i)Sogou\s*(web\s*)?spider`, `(?i)Sogou\s*inst\s*spider`},
			DNSDomains: []string{"sogou.com"},
		},
		"360": {
			Name:       "360蜘蛛",
			Enabled:    true,
			Patterns:   []string{`(?i)360Spider`, `(?i)HaosouSpider`, `(?i)360JK`},
			DNSDomains: []string{"360.cn", "so.com"},
		},
		"shenma": {
			Name:       "神马蜘蛛",
			Enabled:    true,
			Patterns:   []string{`(?i)YisouSpider`, `(?i)Yisouspider`},
			DNSDomains: []string{"sm.cn"},
		},
		"toutiao": {
			Name:       "头条蜘蛛",
			Enabled:    true,
			Patterns:   []string{`(?i)Bytespider`, `(?i)Bytedance`},
			DNSDomains: []string{"bytedance.com"},
		},
		"yandex": {
			Name:       "Yandex蜘蛛",
			Enabled:    true,
			Patterns:   []string{`(?i)YandexBot`, `(?i)YandexImages`, `(?i)YandexMobileBot`},
			DNSDomains: []string{"yandex.ru", "yandex.com", "yandex.net"},
		},
	}
}

// Global spider detector instance
var globalSpiderDetector *SpiderDetector
var spiderDetectorOnce sync.Once
var spiderDetectorMu sync.RWMutex
var globalConfigPath string

// SetSpiderConfigPath sets the global config path for spider detector
// Must be called before GetSpiderDetector() for the first time
func SetSpiderConfigPath(configPath string) {
	spiderDetectorMu.Lock()
	defer spiderDetectorMu.Unlock()
	globalConfigPath = configPath
}

// GetSpiderDetector returns the global spider detector instance
func GetSpiderDetector() *SpiderDetector {
	spiderDetectorOnce.Do(func() {
		spiderDetectorMu.RLock()
		configPath := globalConfigPath
		spiderDetectorMu.RUnlock()

		if configPath == "" {
			configPath = DefaultSpiderConfigPath()
		}

		var err error
		globalSpiderDetector, err = NewSpiderDetectorWithConfig(configPath)
		if err != nil {
			log.Warn().Err(err).Str("path", configPath).Msg("Failed to load spider config, using fallback")
			globalSpiderDetector = newSpiderDetectorFallback()
		}
	})
	return globalSpiderDetector
}

// ResetSpiderDetector resets the global spider detector (mainly for testing)
func ResetSpiderDetector() {
	spiderDetectorMu.Lock()
	defer spiderDetectorMu.Unlock()

	if globalSpiderDetector != nil {
		globalSpiderDetector.StopWatching()
	}
	globalSpiderDetector = nil
	spiderDetectorOnce = sync.Once{}
	globalConfigPath = ""
}

File: content_worker/core/spider_detector.py
"""
蜘蛛检测器模块

检测搜索引擎爬虫，支持User-Agent匹配和DNS反向验证。

主要功能:
- detect_spider(): 检测是否为蜘蛛
- verify_spider_dns(): DNS反向验证
- log_spider_visit(): 记录蜘蛛访问日志

支持的蜘蛛:
- 百度 (Baiduspider)
- 谷歌 (Googlebot)
- 必应 (Bingbot)
- 搜狗 (Sogou)
- 360 (360Spider)
- 神马 (YisouSpider)
- 头条 (Bytespider)
"""
import re
import socket
import asyncio
from typing import Any, Dict, List, NamedTuple, Optional, Tuple
from dataclasses import dataclass
from datetime import datetime
from functools import lru_cache

from loguru import logger


class SpiderInfo(NamedTuple):
    """蜘蛛信息"""
    type: str           # 蜘蛛类型标识
    name: str           # 蜘蛛名称
    dns_domains: List[str]  # DNS验证域名后缀


# 蜘蛛配置
SPIDER_CONFIG: Dict[str, SpiderInfo] = {
    'baidu': SpiderInfo(
        type='baidu',
        name='百度蜘蛛',
        dns_domains=['baidu.com', 'baidu.jp']
    ),
    'google': SpiderInfo(
        type='google',
        name='谷歌蜘蛛',
        dns_domains=['googlebot.com', 'google.com']
    ),
    'bing': SpiderInfo(
        type='bing',
        name='必应蜘蛛',
        dns_domains=['search.msn.com']
    ),
    'sogou': SpiderInfo(
        type='sogou',
        name='搜狗蜘蛛',
        dns_domains=['sogou.com']
    ),
    '360': SpiderInfo(
        type='360',
        name='360蜘蛛',
        dns_domains=['360.cn', 'so.com']
    ),
    'shenma': SpiderInfo(
        type='shenma',
        name='神马蜘蛛',
        dns_domains=['sm.cn']
    ),
    'toutiao': SpiderInfo(
        type='toutiao',
        name='头条蜘蛛',
        dns_domains=['bytedance.com']
    ),
    'yandex': SpiderInfo(
        type='yandex',
        name='Yandex蜘蛛',
        dns_domains=['yandex.ru', 'yandex.com', 'yandex.net']
    ),
}

# User-Agent匹配规则
UA_PATTERNS: List[Tuple[str, re.Pattern]] = [
    ('baidu', re.compile(r'Baiduspider|Baidu-YunGuanCe', re.I)),
    ('google', re.compile(r'Googlebot|Google-InspectionTool|Mediapartners-Google', re.I)),
    ('bing', re.compile(r'bingbot|msnbot|BingPreview', re.I)),
    ('sogou', re.compile(r'Sogou\s*(web\s*)?spider|Sogou\s*inst\s*spider', re.I)),
    ('360', re.compile(r'360Spider|HaosouSpider|360JK', re.I)),
    ('shenma', re.compile(r'YisouSpider|Yisouspider', re.I)),
    ('toutiao', re.compile(r'Bytespider|Bytedance', re.I)),
    ('yandex', re.compile(r'YandexBot|YandexImages|YandexMobileBot', re.I)),
]


@dataclass
class DetectionResult:
    """检测结果"""
    is_spider: bool
    spider_type: Optional[str] = None
    spider_name: Optional[str] = None
    dns_verified: bool = False
    ip: Optional[str] = None
    user_agent: Optional[str] = None


class SpiderDetector:
    """
    蜘蛛检测器

    通过User-Agent和DNS反向验证识别搜索引擎爬虫。

    Attributes:
        enable_dns_verify: 是否启用DNS验证
        dns_verify_types: 需要DNS验证的蜘蛛类型列表
        cache_ttl: DNS缓存时间（秒）

    Example:
        >>> detector = SpiderDetector()
        >>> result = detector.detect("Baiduspider", "220.181.108.95")
        >>> print(result.is_spider, result.spider_name)
        True 百度蜘蛛
    """

    def __init__(
        self,
        enable_dns_verify: bool = True,
        dns_verify_types: Optional[List[str]] = None,
        dns_timeout: float = 2.0
    ):
        """
        初始化检测器

        Args:
            enable_dns_verify: 是否启用DNS反向验证
            dns_verify_types: 需要DNS验证的蜘蛛类型
            dns_timeout: DNS查询超时时间（秒）
        """
        self.enable_dns_verify = enable_dns_verify
        self.dns_verify_types = dns_verify_types or ['baidu', 'google', 'bing']
        self.dns_timeout = dns_timeout
        self._dns_cache: Dict[str, Tuple[str, datetime]] = {}

    def detect(
        self,
        user_agent: str,
        ip: Optional[str] = None
    ) -> DetectionResult:
        """
        检测是否为蜘蛛（同步版本）

        Args:
            user_agent: User-Agent字符串
            ip: 客户端IP地址

        Returns:
            DetectionResult检测结果
        """
        # UA匹配
        spider_type = self._match_user_agent(user_agent)

        if not spider_type:
            return DetectionResult(
                is_spider=False,
                ip=ip,
                user_agent=user_agent
            )

        spider_info = SPIDER_CONFIG.get(spider_type)

        # DNS验证
        dns_verified = False
        if (
            self.enable_dns_verify
            and ip
            and spider_type in self.dns_verify_types
        ):
            dns_verified = self._verify_dns_sync(ip, spider_info.dns_domains)

        return DetectionResult(
            is_spider=True,
            spider_type=spider_type,
            spider_name=spider_info.name if spider_info else spider_type,
            dns_verified=dns_verified,
            ip=ip,
            user_agent=user_agent
        )

    async def detect_async(
        self,
        user_agent: str
    ) -> DetectionResult:
        """
        检测是否为蜘蛛（异步版本，仅UA匹配）

        Args:
            user_agent: User-Agent字符串

        Returns:
            DetectionResult检测结果
        """
        # UA匹配
        spider_type = self._match_user_agent(user_agent)

        if not spider_type:
            return DetectionResult(
                is_spider=False,
                user_agent=user_agent
            )

        spider_info = SPIDER_CONFIG.get(spider_type)

        return DetectionResult(
            is_spider=True,
            spider_type=spider_type,
            spider_name=spider_info.name if spider_info else spider_type,
            user_agent=user_agent
        )

    @lru_cache(maxsize=1000)
    def _match_user_agent(self, user_agent: str) -> Optional[str]:
        """
        匹配User-Agent

        Args:
            user_agent: UA字符串

        Returns:
            匹配的蜘蛛类型或None
        """
        if not user_agent:
            return None

        for spider_type, pattern in UA_PATTERNS:
            if pattern.search(user_agent):
                return spider_type

        return None

    def _get_cached_hostname(self, ip: str) -> Optional[str]:
        """从缓存获取主机名，过期返回None"""
        if ip not in self._dns_cache:
            return None
        hostname, cached_time = self._dns_cache[ip]
        if (datetime.now() - cached_time).total_seconds() >= 3600:
            return None
        return hostname

    def _matches_valid_domain(self, hostname: str, valid_domains: List[str]) -> bool:
        """检查主机名是否匹配有效域名后缀"""
        return any(hostname.endswith(d) for d in valid_domains)

    def _verify_dns_sync(
        self,
        ip: str,
        valid_domains: List[str]
    ) -> bool:
        """
        DNS反向验证（同步）

        Args:
            ip: IP地址
            valid_domains: 有效的域名后缀列表

        Returns:
            是否验证通过
        """
        try:
            # 检查缓存
            cached_hostname = self._get_cached_hostname(ip)
            if cached_hostname:
                return self._matches_valid_domain(cached_hostname, valid_domains)

            # 反向DNS查询
            hostname, _, _ = socket.gethostbyaddr(ip)

            # 缓存结果
            self._dns_cache[ip] = (hostname, datetime.now())

            return self._matches_valid_domain(hostname, valid_domains)

        except (socket.herror, socket.gaierror, socket.timeout) as e:
            logger.debug(f"DNS verification failed for {ip}: {e}")
            return False

    async def _verify_dns_async(
        self,
        ip: str,
        valid_domains: List[str]
    ) -> bool:
        """
        DNS反向验证（异步）

        Args:
            ip: IP地址
            valid_domains: 有效的域名后缀列表

        Returns:
            是否验证通过
        """
        try:
            # 检查缓存
            cached_hostname = self._get_cached_hostname(ip)
            if cached_hostname:
                return self._matches_valid_domain(cached_hostname, valid_domains)

            # 异步反向DNS查询
            loop = asyncio.get_event_loop()
            hostname, _, _ = await asyncio.wait_for(
                loop.run_in_executor(None, socket.gethostbyaddr, ip),
                timeout=self.dns_timeout
            )

            # 缓存结果
            self._dns_cache[ip] = (hostname, datetime.now())

            return self._matches_valid_domain(hostname, valid_domains)

        except (socket.herror, socket.gaierror, socket.timeout, asyncio.TimeoutError) as e:
            logger.debug(f"DNS verification failed for {ip}: {e}")
            return False

    def is_spider(self, user_agent: str) -> bool:
        """
        快速检测是否为蜘蛛（仅UA匹配）

        Args:
            user_agent: UA字符串

        Returns:
            是否为蜘蛛
        """
        return self._match_user_agent(user_agent) is not None

    def get_spider_type(self, user_agent: str) -> Optional[str]:
        """
        获取蜘蛛类型

        Args:
            user_agent: UA字符串

        Returns:
            蜘蛛类型或None
        """
        return self._match_user_agent(user_agent)

    def clear_cache(self) -> None:
        """清空DNS缓存"""
        self._dns_cache.clear()
        self._match_user_agent.cache_clear()

    def get_stats(self) -> Dict[str, Any]:
        """获取缓存统计"""
        return {
            'dns_cache_size': len(self._dns_cache),
            'ua_cache_info': self._match_user_agent.cache_info()._asdict()
        }


# 全局检测器实例
_detector: Optional[SpiderDetector] = None


def get_spider_detector() -> SpiderDetector:
    """获取全局蜘蛛检测器"""
    global _detector
    if _detector is None:
        _detector = SpiderDetector()
    return _detector


def init_spider_detector(
    enable_dns_verify: bool = True,
    dns_verify_types: Optional[List[str]] = None,
    dns_timeout: float = 2.0
) -> SpiderDetector:
    """
    初始化全局蜘蛛检测器

    Args:
        enable_dns_verify: 是否启用DNS验证
        dns_verify_types: 需要DNS验证的类型
        dns_timeout: DNS超时时间

    Returns:
        SpiderDetector实例
    """
    global _detector
    _detector = SpiderDetector(
        enable_dns_verify=enable_dns_verify,
        dns_verify_types=dns_verify_types,
        dns_timeout=dns_timeout
    )
    return _detector


def is_spider(user_agent: str) -> bool:
    """
    快捷函数 - 检测是否为蜘蛛

    Args:
        user_agent: UA字符串

    Returns:
        是否为蜘蛛
    """
    return get_spider_detector().is_spider(user_agent)


def detect_spider(
    user_agent: str,
    ip: Optional[str] = None
) -> DetectionResult:
    """
    快捷函数 - 完整检测

    Args:
        user_agent: UA字符串
        ip: IP地址

    Returns:
        DetectionResult
    """
    return get_spider_detector().detect(user_agent, ip)


async def detect_spider_async(user_agent: str) -> DetectionResult:
    """
    快捷函数 - 异步检测（仅UA匹配）

    Args:
        user_agent: UA字符串

    Returns:
        DetectionResult
    """
    return await get_spider_detector().detect_async(user_agent)
